File: decision/usecases/decision_service.py
"""Use case for making decisions."""

from domain.models import (
    DecisionRequest, 
    DecisionResult, 
    DecisionConfigModel,
    DecisionType
)
from domain.repository import DecisionConfigRepository


class DecisionService:
    """Service for making fraud detection decisions."""
    
    def __init__(self, config_repository: DecisionConfigRepository):
        self.config_repository = config_repository
    
    def make_decision(self, request: DecisionRequest) -> DecisionResult:
        """Make a decision based on aggregated engine scores."""
        # Get current configuration
        config = self.config_repository.get_config()
        
        # Calculate final weighted score
        final_score = self._calculate_final_score(
            config,
            request.rule_engine_score,
            request.anomaly_detection_score,
            request.predictive_engine_score
        )
        
        # Convert to percentage
        final_score_percent = final_score * 100
        
        # Determine decision
        decision = self._get_decision(final_score_percent, config)
        
        # Calculate confidence
        confidence = self._calculate_confidence(decision, final_score_percent, config)
        
        # Create breakdown
        breakdown = self._create_breakdown(
            config,
            request.rule_engine_score,
            request.anomaly_detection_score,
            request.predictive_engine_score
        )
        
        return DecisionResult(
            entity_id=request.entity_id,
            final_score=final_score,
            final_score_percent=final_score_percent,
            decision=decision,
            breakdown=breakdown,
            confidence=confidence
        )
    
    def _calculate_final_score(
        self, 
        config: DecisionConfigModel,
        rule_score: float, 
        anomaly_score: float, 
        predictive_score: float
    ) -> float:
        """Calculate final weighted score."""
        if config.model_based_scoring:
            # Future: Use ML model to determine weights
            # For now, fall back to manual weights
            pass
        
        # Calculate weighted average
        weighted_sum = (
            rule_score * config.rule_engine_weight +
            anomaly_score * config.anomaly_detection_weight +
            predictive_score * config.predictive_engine_weight
        )
        final_score = weighted_sum / 100.0
        
        return final_score
    
    def _get_decision(
        self, 
        score_percent: float, 
        config: DecisionConfigModel
    ) -> str:
        """Determine decision based on thresholds."""
        if score_percent <= config.auto_approve_threshold:
            decision = DecisionType.AUTO_APPROVE
        elif score_percent >= config.auto_reject_threshold:
            decision = DecisionType.AUTO_REJECT
        else:
            decision = DecisionType.HUMAN_REVIEW
        
        return decision
    
    def _calculate_confidence(
        self, 
        decision: str, 
        score_percent: float, 
        config: DecisionConfigModel
    ) -> float:
        """Calculate decision confidence."""
        if decision == DecisionType.AUTO_APPROVE:
            confidence = 1.0 - (score_percent / config.auto_approve_threshold)
        elif decision == DecisionType.AUTO_REJECT:
            confidence = 1.0 - ((100 - score_percent) / (100 - config.auto_reject_threshold))
        else:
            confidence = 0.5
        
        return max(0.0, min(1.0, confidence))
    
    def _create_breakdown(
        self,
        config: DecisionConfigModel,
        rule_score: float,
        anomaly_score: float,
        predictive_score: float
    ) -> dict:
        """Create score breakdown by engine."""
        return {
            "rule_engine": {
                "score": rule_score,
                "weight": config.rule_engine_weight,
                "contribution": rule_score * config.rule_engine_weight
            },
            "anomaly_detection": {
                "score": anomaly_score,
                "weight": config.anomaly_detection_weight,
                "contribution": anomaly_score * config.anomaly_detection_weight
            },
            "predictive_engine": {
                "score": predictive_score,
                "weight": config.predictive_engine_weight,
                "contribution": predictive_score * config.predictive_engine_weight
            }
        }


class ConfigService:
    """Service for managing configuration."""
    
    def __init__(self, config_repository: DecisionConfigRepository):
        self.config_repository = config_repository
    
    def get_config(self) -> DecisionConfigModel:
        """Get current configuration."""
        return self.config_repository.get_config()
    
    def update_config(self, config: DecisionConfigModel) -> dict:
        """Update configuration."""
        return self.config_repository.update_config(config)

File: backend/internal/adapter/client/decisionservice.go
package client

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// DecisionServiceClient handles communication with the decision service
type DecisionServiceClient struct {
	baseURL    string
	httpClient *http.Client
}

// NewDecisionServiceClient creates a new decision service client
func NewDecisionServiceClient(baseURL string) *DecisionServiceClient {
	return &DecisionServiceClient{
		baseURL: baseURL,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// DecisionConfigModel represents the decision service configuration
type DecisionConfigModel struct {
	ModelBasedThresholds   bool    `json:"model_based_thresholds"`
	AutoApproveThreshold   float64 `json:"auto_approve_threshold"`
	AutoRejectThreshold    float64 `json:"auto_reject_threshold"`
	ModelBasedScoring      bool    `json:"model_based_scoring"`
	RuleEngineWeight       float64 `json:"rule_engine_weight"`
	AnomalyDetectionWeight float64 `json:"anomaly_detection_weight"`
	PredictiveEngineWeight float64 `json:"predictive_engine_weight"`
}

// DecisionRequest represents a request for a decision
type DecisionRequest struct {
	EntityID              string  `json:"entity_id"`
	RuleEngineScore       float64 `json:"rule_engine_score"`
	AnomalyDetectionScore float64 `json:"anomaly_detection_score"`
	PredictiveEngineScore float64 `json:"predictive_engine_score"`
	DataType              string  `json:"data_type,omitempty"`
}

// ScoreBreakdown represents the breakdown of scores by engine
type ScoreBreakdown struct {
	RuleEngine struct {
		Score        float64 `json:"score"`
		Weight       float64 `json:"weight"`
		Contribution float64 `json:"contribution"`
	} `json:"rule_engine"`
	AnomalyDetection struct {
		Score        float64 `json:"score"`
		Weight       float64 `json:"weight"`
		Contribution float64 `json:"contribution"`
	} `json:"anomaly_detection"`
	PredictiveEngine struct {
		Score        float64 `json:"score"`
		Weight       float64 `json:"weight"`
		Contribution float64 `json:"contribution"`
	} `json:"predictive_engine"`
}

// DecisionResponse represents the response from decision service
type DecisionResponse struct {
	EntityID          string         `json:"entity_id"`
	FinalScore        float64        `json:"final_score"`
	FinalScorePercent float64        `json:"final_score_percent"`
	Decision          string         `json:"decision"`
	Breakdown         ScoreBreakdown `json:"breakdown"`
	Confidence        float64        `json:"confidence"`
}

// ConfigUpdateResponse represents the response from config update
type ConfigUpdateResponse struct {
	Success bool                `json:"success"`
	Message string              `json:"message"`
	Config  DecisionConfigModel `json:"config"`
}

// GetConfig retrieves the current decision service configuration
func (c *DecisionServiceClient) GetConfig() (*DecisionConfigModel, error) {
	url := c.baseURL + "/config"

	httpReq, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("decision service returned status %d", resp.StatusCode)
	}

	var config DecisionConfigModel
	if err := json.NewDecoder(resp.Body).Decode(&config); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &config, nil
}

// UpdateConfig updates the decision service configuration
func (c *DecisionServiceClient) UpdateConfig(config DecisionConfigModel) (*DecisionConfigModel, error) {
	url := c.baseURL + "/config"

	jsonData, err := json.Marshal(config)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("decision service returned status %d", resp.StatusCode)
	}

	var result ConfigUpdateResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !result.Success {
		return nil, fmt.Errorf("failed to update config: %s", result.Message)
	}

	return &result.Config, nil
}

// MakeDecision makes a decision based on scores from multiple engines
func (c *DecisionServiceClient) MakeDecision(entityID string, ruleEngineScore, anomalyDetectionScore, predictiveEngineScore float64, dataType string) (*DecisionResponse, error) {
	req := DecisionRequest{
		EntityID:              entityID,
		RuleEngineScore:       ruleEngineScore,
		AnomalyDetectionScore: anomalyDetectionScore,
		PredictiveEngineScore: predictiveEngineScore,
		DataType:              dataType,
	}
	return c.decide("/decide", req)
}

// decide makes a request to the decision service
func (c *DecisionServiceClient) decide(endpoint string, req DecisionRequest) (*DecisionResponse, error) {
	url := c.baseURL + endpoint

	jsonData, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("decision service returned status %d", resp.StatusCode)
	}

	var response DecisionResponse
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &response, nil
}

// MergedDataTypeConfig represents merged configuration for a data type
type MergedDataTypeConfig struct {
	DataType         string              `json:"data_type"`
	Config           DecisionConfigModel `json:"config"`
	IsCustom         bool                `json:"is_custom"`
	OverriddenFields []string            `json:"overridden_fields"`
}

// DataTypeDecisionConfig represents data-type-specific config overrides
type DataTypeDecisionConfig struct {
	DataType               string   `json:"data_type"`
	AutoApproveThreshold   *float64 `json:"auto_approve_threshold,omitempty"`
	AutoRejectThreshold    *float64 `json:"auto_reject_threshold,omitempty"`
	RuleEngineWeight       *float64 `json:"rule_engine_weight,omitempty"`
	AnomalyDetectionWeight *float64 `json:"anomaly_detection_weight,omitempty"`
	PredictiveEngineWeight *float64 `json:"predictive_engine_weight,omitempty"`
	ModelBasedScoring      *bool    `json:"model_based_scoring,omitempty"`
	ModelBasedThresholds   *bool    `json:"model_based_thresholds,omitempty"`
}

// GetDataTypeConfig retrieves merged configuration for a specific data type
func (c *DecisionServiceClient) GetDataTypeConfig(dataType string) (*MergedDataTypeConfig, error) {
	url := fmt.Sprintf("%s/config/data-types/%s", c.baseURL, dataType)

	httpReq, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("decision service returned status %d", resp.StatusCode)
	}

	var config MergedDataTypeConfig
	if err := json.NewDecoder(resp.Body).Decode(&config); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &config, nil
}

// UpdateDataTypeConfig updates data-type-specific configuration
func (c *DecisionServiceClient) UpdateDataTypeConfig(dataType string, config DataTypeDecisionConfig) (*DataTypeDecisionConfig, error) {
	url := fmt.Sprintf("%s/config/data-types/%s", c.baseURL, dataType)

	jsonData, err := json.Marshal(config)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("decision service returned status %d", resp.StatusCode)
	}

	var result DataTypeDecisionConfig
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

// GetAllDataTypeConfigs retrieves all data types with custom configs
func (c *DecisionServiceClient) GetAllDataTypeConfigs() ([]string, error) {
	url := c.baseURL + "/config/data-types"

	httpReq, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("decision service returned status %d", resp.StatusCode)
	}

	var result struct {
		DataTypes []map[string]string `json:"data_types"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	dataTypes := make([]string, len(result.DataTypes))
	for i, dt := range result.DataTypes {
		dataTypes[i] = dt["data_type"]
	}

	return dataTypes, nil
}

// DeleteDataTypeConfig deletes data-type-specific configuration
func (c *DecisionServiceClient) DeleteDataTypeConfig(dataType string) error {
	url := fmt.Sprintf("%s/config/data-types/%s", c.baseURL, dataType)

	httpReq, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("decision service returned status %d", resp.StatusCode)
	}

	return nil
}

// HealthCheck checks if the decision service is healthy
func (c *DecisionServiceClient) HealthCheck() error {
	resp, err := c.httpClient.Get(c.baseURL + "/health")
	if err != nil {
		return fmt.Errorf("failed to check health: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("decision service health check failed with status %d", resp.StatusCode)
	}

	return nil
}
